### PR TITLE
8313718: make container at requires command configurable

### PR DESCRIPTION
**Pull request by mseledts**
Container ecosystem is growing. It would be beneficial to define custom command to figure out whether a specific test host or environment allows for container testing. This enhancement seeks to make the command used by jtreg "requires" extension configurable, specifically test/jtreg-ext/requires/VMProps.java checkContainerSupport().

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313718](https://bugs.openjdk.org/browse/JDK-8313718): make container at requires command configurable (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15475/head:pull/15475` \
`$ git checkout pull/15475`

Update a local copy of the PR: \
`$ git checkout pull/15475` \
`$ git pull https://git.openjdk.org/jdk.git pull/15475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15475`

View PR using the GUI difftool: \
`$ git pr show -t 15475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15475.diff">https://git.openjdk.org/jdk/pull/15475.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15475#issuecomment-1698210670)

### CODE REVIEW

File: test/jtreg-ext/requires/VMProps.java [hunk 1]
```diff
@@ -527,16 +528,16 @@ protected String isCompiler2Enabled() {
...
-     * @return true if docker is supported in a given environment
+     * @return true if container subsystem is supported in a given environment
      */
-    protected String dockerSupport() {
-        log("Entering dockerSupport()");
+    protected String containerSupport() {
+        log("Entering containerSupport()");
 
-        boolean isSupported = false;
+        boolean isSupported = true;
```
Comment (dholmes-ora) [inline test/jtreg-ext/requires/VMProps.java:538]: This needs to be initialized to false.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
I'm unclear how this is supposed to work. We have existing tests that have `@requires docker.support` that may or may not work for arbitrary "container" environments. What if they are really docker specific tests? Won't we still need `dockerSupport`?

Thanks

File: test/jtreg-ext/requires/VMProps.java (hunk 1 above)
Comment (mseledts) [reply test/jtreg-ext/requires/VMProps.java:538]: Good catch, thank you. This is a bug, a leftover from my experimentation.

**Comment by mseledts** [comment]
Hi David,

Thank you for taking a look at this change.  To answer your questions:
 -- how is this supposed to work ? 
 If user running the tests does not specify the newly introduced property things will work same as before. If user specifies -Djdk.test.container.requires.check.command then the specified command will be executed to check whether container testing can be performed on a given host/system. 

By default the code in VMProps.java runs "docker ps". With this change we could override it with any other relevant command, such as "podman version" or other type of container command.

-- What if they are really docker specific tests?
I think our container tests should be fairly container-generic (or container agnostic), but need to check this. This raises a good question. Perhaps we should consider removing the at-requires check entirely, and rely on individual checks in individual tests. This will give us more flexibility, we run a specific check in each test and throw a jtreg.SkippedException upon failure. We already do some additional checks in each test, such as podman version, cgroups v1 vs v2, etc.

Let me know what you think.

**Comment by dholmes-ora** [comment]
@mseledts  My concern is that given:
```
map.put("docker.support", this::containerSupport)
```
we will now have a situation where `containerSupport` may indicate `podman` but that will enable `requires docker.support`. It seems to me that we need a range of potential checks where the generic `requires container.support` is for a test targeted to any container environment (though I wonder in that case what exactly it requires that the test depends on???); and we would then potentially need separate `docker.support`, `podman.support` if tests did things specific to those container environments.

**Comment by mseledts** [comment]
David,
  We already have a way to configure engine executable, that is 'docker' vs 'podman' vs 'some-other-engine'. However the command that we use to test whether the engine is available and working on a given test host is "<CONTAINER_ENGINE> ps". However the "ps" sub-command may not be applicable for all container types, and it may be "heavy" at times on some busy hosts.
  One option is to replace it with something more universal and light-weight, such as "<CONTAINER_ENGINE> help". The "help" command should be available for all containers, and be reasonably light-weight. 
  The other option is to make it fully configurable as this change. 
It would also be good to have an ability to skip such check entirely for hosts know to be properly configured.

**Pull request closed without merging**